### PR TITLE
Discard GC data when loading state from snapshot with different GC version than current

### PR DESCRIPTION
**Pull request by agarwal-navin**
When GC loads from a snapshot or updates latest state from a snapshot (in refreshLatestSummary), if the GC version of the snapshot is different from the current GC version, reset all local GC state and only update deleted nodes from this snapshot.

When the GC version of a snapshot is different from the current, the GC data in the snapshot cannot be trusted. The GC data has to be re-generated (in the next GC run) because the previous GC data was generated by a version that may have issues - this already happens today. The same logic should apply when updating the garbage collector's state from the snapshot. Note that we still update the deleted nodes state because these nodes have been deleted and cannot be brought back.

[AB#3779](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3779)

### CODE REVIEW

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 1]
```diff
@@ -38,13 +41,22 @@ import {
...
-import { dataStoreAttributesBlobName, RefreshSummaryResult } from "../../summary";
+import {
+	dataStoreAttributesBlobName,
+	IContainerRuntimeMetadata,
+	metadataBlobName,
+	RefreshSummaryResult,
+} from "../../summary";
 import { pkgVersion } from "../../packageVersion";
+// eslint-disable-next-line import/no-internal-modules
+import { IGCSummaryTrackingData } from "../../gc/gcSummaryStateTracker";
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:59]: Why not add this to gc/index.ts?

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 2]
```diff
@@ -124,14 +134,27 @@ describe("Garbage Collection Tests", () => {
...
+
 		return GarbageCollector.create({
 			...createParams,
 			runtime: gcRuntime,
 			gcOptions: createParams.gcOptions ?? {},
 			baseSnapshot: createParams.baseSnapshot,
 			baseLogger: mc.logger,
-			existing: createParams.metadata !== undefined /* existing */,
-			metadata: createParams.metadata,
+			existing: createParams.baseSnapshot !== undefined /* existing */,
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:156]:
```suggestion
			existing,
```

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 3]
```diff
@@ -213,6 +241,7 @@ describe("Garbage Collection Tests", () => {
 
 		const summarizerContainerTests = (
 			timeout: number,
+			mode: "inactive" | "sweep",
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:244]: Future - could remove the following fooEventName params and compute them from this mode.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 4]
```diff
@@ -696,13 +730,302 @@ describe("Garbage Collection Tests", () => {
...
+				"sweep",
 				"GarbageCollector:SweepReadyObject_Revived",
 				"GarbageCollector:SweepReadyObject_Changed",
 				"GarbageCollector:SweepReadyObject_Loaded",
 				false, // expectDeleteLogs
 			);
 		});
 
+		describe("GC version changes", () => {
+			type GcWithStateOverride = IGarbageCollector & {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:742]: Why not use the GcWithPrivates type (and add the stuff you need)

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 5]
```diff
@@ -696,13 +730,302 @@ describe("Garbage Collection Tests", () => {
...
+					gcBlobsMap,
+					gcMetadata,
+				) as GcWithStateOverride;
+			};
+
+			function getSnapshotWithGCVersion(gcVersion: GCVersion) {
+				// Create a snapshot tree to be used as the GC snapshot tree.
+				const gcSnapshotTree = getDummySnapshotTree();
+				const gcBlobId = "root";
+				// Add a GC blob with key that start with `gcBlobPrefix` to the GC snapshot tree. The blob Id for this
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:773]:
```suggestion
				// Add a GC blob with key that start with `blobs` to the GC snapshot tree. The blob Id for this
```

`gcBlobPrefix` is the variable name not the actual value, right?

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 6]
```diff
@@ -696,13 +730,302 @@ describe("Garbage Collection Tests", () => {
...
+				// Create GC state with a node. This will be returned when the garbage collector asks for the GC blob
+				// with `gcBlobId`.
+				const gcState: IGarbageCollectionState = { gcNodes: {} };
+				const nodeData: IGarbageCollectionNodeData = {
+					outboundRoutes: [],
+					unreferencedTimestampMs: 123,
+				};
+				gcState.gcNodes[nodes[0]] = nodeData;
+
+				// Create a tombstone blob a node as tombstone. This will be returned when the garbage collector asks for
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:786]: Some typo here

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 7]
```diff
@@ -696,13 +730,302 @@ describe("Garbage Collection Tests", () => {
...
+				// the tombstone blob.
+				const gcTombstoneBlobId = "tombstone";
+				gcSnapshotTree.blobs[gcTombstoneBlobKey] = gcTombstoneBlobId;
+				const tombstones = [nodes[0]];
+
+				// Create a tombstone blob a node as tombstone. This will be returned when the garbage collector asks for
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:792]: Should be about deletedNodes not tombstones

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 8]
```diff
@@ -696,13 +730,302 @@ describe("Garbage Collection Tests", () => {
...
+				// the tombstone blob.
+				const gcDeletedBlobId = "deletedNodes";
+				gcSnapshotTree.blobs[gcDeletedBlobKey] = gcDeletedBlobId;
+				const deletedBlobs = [nodes[0]];
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:796]: Is it intentional that you're putting the same node in both tombstone and deleted list? I guess that's what will happen for a document that's had tombstone enabled and then switches to Sweep?

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 9]
```diff
@@ -696,13 +730,302 @@ describe("Garbage Collection Tests", () => {
...
+				// Initialize from the base state and validate that tombstones and deleted state both have one entry
+				// as per the base snapshot.
+				await garbageCollector.initializeBaseState();
+				assert.strictEqual(garbageCollector.tombstones.length, 1);
+				assert.strictEqual(garbageCollector.deletedNodes.size, 1);
+			});
+
+			it("discards GC state and tombstone state in base snapshot when GC version changes", async () => {
+				const updatedGCVersion = currentGCVersion + 1;
+				const { baseSnapshot, gcBlobMap } = getSnapshotWithGCVersion(updatedGCVersion);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:855]: Do you want to move this to within `createGCOverride`, so then you just pass the gcVersion and it creates the snapshot/blob map and creates the GC?  Otherwise you could accidentally pass different gcVersions for the snapshot contents and the metadata.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts [hunk 10]
```diff
@@ -696,13 +730,302 @@ describe("Garbage Collection Tests", () => {
...
+				const { baseSnapshot, gcBlobMap } = getSnapshotWithGCVersion(currentGCVersion);
+				const garbageCollector = createGCOverride(
+					baseSnapshot,
+					gcBlobMap,
+					currentGCVersion,
+				);
+
+				// GC state, tombstone state and deleted nodes should all be read from base snapshot.
+				const baseSnapshotData = await garbageCollector.baseSnapshotDataP;
+				assert(
```
Comment (markfields) [inline packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:898]: I wouldn't bother asserting these since it's redundant with the other test. Or maybe pile them into one assert with message "PRECONDITION FAILED" or something.

I wish Mocha had a `precondition` module with `precondition.assert` functions to match `assert` functions but if they failed, it doesn't fail the test the same as if a regular assert fails.  Would help reduce investigation surface area when something fundamental breaks a million tests, ya know?

File: packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts [hunk 11]
```diff
@@ -276,38 +290,38 @@ export class GCSummaryStateTracker {
...
-			this.latestSummaryData = undefined;
-			return;
+		let snapshotData = await getGCDataFromSnapshot(gcSnapshotTree, readAndParseBlob);
+
+		// If the GC version in the snapshot does not match the GC version currently in effect, the GC data
+		// in the snapshot cannot be interpreted correctly. Set everything to undefined except for deletedNodes
+		// because irrespective of GC versions, these nodes have been deleted and cannot be brought back. The
+		// deletedNodes info is needed to identify when these nodes are used.
+		if (getGCVersion(metadata) !== this.currentGCVersion) {
+			snapshotData = {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:313]: @markfields I was thinking that if refreshing from snapshot, we don't need to update deleted nodes right. We need to do it when loading from a base snapshot but when refreshing, we already have deleted nodes from whatever the last GC run was and we can continue using that.
The only problem with updating it here is that if we move back in time, the deleted nodes unncessarily move back in time too.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 8 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:796]: That wasn't the intention. These tests don't care about the contents of tombstone / deleted nodes. They just care that they exist.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 9 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:855]: That's a good point.

File: packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts (hunk 11 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:313]: As discussed elsewhere, makes sense to me, since it's not used for incremental summary calculation. We can keep that in mind when we revisit the state machine and different rolls the different in-memory things have.

File: packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts [hunk 12]
```diff
@@ -276,38 +290,38 @@ export class GCSummaryStateTracker {
 		// If the summary was not tracked by this client, the state should be updated from the downloaded snapshot.
 		const snapshotTree = result.snapshotTree;
 		const metadataBlobId = snapshotTree.blobs[metadataBlobName];
-		if (metadataBlobId) {
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:279]: Are there cases where metadata is missing...?  i.e. can we assert it's not `undefined`?

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:279]: I don't think so. It was added a long time back and there shouldn't be active clients that generate summary without metadata.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:59]: I was being lazy :-). Will add it.

File: packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/test/gc/garbageCollection.spec.ts:773]: yeah, its not. will update.

File: packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts [hunk 13]
```diff
@@ -276,38 +290,38 @@ export class GCSummaryStateTracker {
...
+				deletedNodes: snapshotData.deletedNodes,
+			};
 		}
 
-		// If tracking state across summaries, update latest summary data from the snapshot's GC data.
 		this.latestSummaryData = {
-			serializedGCState: JSON.stringify(generateSortedGCState(gcSnapshotData.gcState)),
-			serializedTombstones: JSON.stringify(gcSnapshotData.tombstones),
-			serializedDeletedNodes: JSON.stringify(gcSnapshotData.deletedNodes),
+			serializedGCState: JSON.stringify(snapshotData.gcState),
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:321]: Is this ok to be `undefined`?

Comment (markfields) [reply packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:321]: Seems like it should be, and `tombstones` and `deletedNodes` already can be `undefined` so this isn't a new code path I guess.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 14]
```diff
@@ -201,16 +201,36 @@ export class GarbageCollector implements IGarbageCollector {
...
-					// back-compat - Older documents will have the GC blobs in each data store's snapshot tree.
-					return getSnapshotDataFromOldSnapshotFormat(
-						baseSnapshot,
-						createParams.metadata,
+					const snapshotData = await getGCDataFromSnapshot(
+						gcSnapshotTree,
 						readAndParseBlob,
 					);
+
+					if (
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:218]: Subjective nit: I would flip this `if` condition and bring the comment up here, so then the bottom of the function is the "happy path" and exceptional cases (like the back compat one above and this one) are called out earlier on.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts [hunk 15]
```diff
@@ -380,11 +401,8 @@ export class GarbageCollector implements IGarbageCollector {
 			this.runtime.updateTombstonedRoutes(this.tombstones);
 		}
 
-		// Update the summary state tracker's state from this snapshot.
-		this.summaryStateTracker.updateStateFromSnapshotData(snapshotData);
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/garbageCollection.ts:384]: Remind me why this can go away?

Comment (markfields) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:384]: Because `summaryStateTracker.initializeBaseState` is called above I guess...?

File: packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts [hunk 16]
```diff
@@ -276,38 +290,38 @@ export class GCSummaryStateTracker {
...
-		this.pendingSummaryData = undefined;
-		return latestGCData;
-	}
 
-	/**
-	 * Update state from the given snapshot data. This is done during load and during refreshing state from a snapshot.
-	 * @param gcSnapshotData - The GC snapshot data to update state from.
-	 */
-	public updateStateFromSnapshotData(gcSnapshotData: IGarbageCollectionSnapshotData | undefined) {
-		// If there is no snapshot data, it means this snapshot was generated with GC disabled. Unset all GC state.
```
Comment (markfields) [inline packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:300]: Is this case lost for places that used to call `updateStateFromSnapshotData`?

**Review by markfields — APPROVED** [review]
🎂 I think it looks good.  Some of the diffs were hard to follow where you shuffled code around, I left some questions about those changes, we can discuss tomorrow if those questions raise any doubts.  I'm trusting existing test coverage here.

File: packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts (hunk 13 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:321]: Yes, it can be undefined. If it is undefined, in the next summary, GC tree will be re-written into the summary instead of a handle.

File: packages/runtime/container-runtime/src/gc/garbageCollection.ts (hunk 15 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/garbageCollection.ts:384]: When loading from snapshot, this is done explicitly via `summaryStateTracker.initializeBaseState` when initializing. When refreshing from snapshot, this happens in the `refreshLatestSummary` method in the summary state tracker.
Not updating this in `refreshLatestSummary` and doing it in a separate step was not correct because it added the posibility of adding inconsistent state.

File: packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts (hunk 16 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/gc/gcSummaryStateTracker.ts:300]: This happens in `refreshLatestSummary` or in `initializeBaseState` making it more explicit rather than a generic update state method.
